### PR TITLE
update rounding behavior in `calculateSolvencyAfterShort`

### PR DESCRIPTION
**Pull request by dpaiton**
# Resolved Issues
working towards https://github.com/delvtech/hyperdrive-rs/issues/29

# Description
The solvency after short check used `divDown` to convert from base to shares, but the open short function uses `divUp`. Therefore it was possible to have the solvency check say "all good" and then open short reverts. This fixes that problem.

I also added some intermediate checks so that the function fails gracefully in more situations.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity, the corresponding checklist can
be removed.

## [[Reviewer Name]]

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### CODE REVIEW

File: test/utils/HyperdriveUtils.sol [hunk 1]
```diff
@@ -1233,19 +1233,27 @@ library HyperdriveUtils {
...
+            calculateShortGovernanceFee(
+                _shortAmount,
+                _spotPrice,
+                _params.curveFee,
+                _params.governanceLPFee
+            )).divUp(_params.vaultSharePrice);
+        if (shareAmount < tradeFees) {
+            return (0, false);
+        }
+
```
Comment (jalextowle) [inline test/utils/HyperdriveUtils.sol:1250]: Nit: Remove new line.

File: test/utils/HyperdriveUtils.sol [hunk 2]
```diff
@@ -1233,19 +1233,27 @@ library HyperdriveUtils {
...
+        uint256 tradeDelta = shareAmount - tradeFees;
+        if (_params.shareReserves < tradeDelta) {
+            return (0, false);
+        }
+
```
Comment (jalextowle) [inline test/utils/HyperdriveUtils.sol:1255]: Nit: Remove new line

File: test/utils/HyperdriveUtils.sol [hunk 3]
```diff
@@ -1233,19 +1233,27 @@ library HyperdriveUtils {
...
+                _shortAmount,
+                _spotPrice,
+                _params.curveFee,
+                _params.governanceLPFee
+            )).divUp(_params.vaultSharePrice);
+        if (shareAmount < tradeFees) {
+            return (0, false);
+        }
+
+        uint256 tradeDelta = shareAmount - tradeFees;
```
Comment (jalextowle) [inline test/utils/HyperdriveUtils.sol:1251]: Nit: Let’s use “shareReservesDelta” to be consistent.

File: test/utils/HyperdriveUtils.sol [hunk 4]
```diff
@@ -1233,19 +1233,27 @@ library HyperdriveUtils {
...
-                    _spotPrice,
-                    _params.curveFee
-                ) -
-                    calculateShortGovernanceFee(
-                        _shortAmount,
-                        _spotPrice,
-                        _params.curveFee,
-                        _params.governanceLPFee
-                    )).divDown(_params.vaultSharePrice));
+        uint256 tradeFees = (calculateShortCurveFee(
```
Comment (jalextowle) [inline test/utils/HyperdriveUtils.sol:1236]: Let’s use “curveFee” to be consistent.

**Review by jalextowle — APPROVED** [review]
LGTM once the nits are addressed.

File: test/utils/HyperdriveUtils.sol (hunk 4 above)
Comment (dpaiton) [reply test/utils/HyperdriveUtils.sol:1236]: it is both the curveFee and the governanceCurveFee, which are treaded differently [elsewhere](https://github.com/delvtech/hyperdrive/blob/02efffd127f197629641c91d5baaae61b7247968/contracts/src/internal/HyperdriveBase.sol#L777-L789). With that in mind do you st ill want to use "curveFee"?

Comment (dpaiton) [reply test/utils/HyperdriveUtils.sol:1236]: updating to `totalCurveFee` per outside discusison.

**Pull request merged**